### PR TITLE
module extension: add support for canonical name resolution

### PR DESCRIPTION
**Pull request by gabrielschulhof**
Before attempting to load a module, each provided resolver must be given an
opportunity to examine the name of the requested module without actually
loading it so as to canonicalize it, in case a module can be referred to by
multiple names.

Then, modules are loaded and cached by their canonical name.

### CODE REVIEW

File: docs/12.EXT-REFERENCE-MODULE.md [hunk 1]
```diff
@@ -87,23 +104,39 @@ typedef jerry_value_t (*jerryx_native_module_on_resolve_t) (void);
...
 **Prototype**
 
 ```c
-typedef bool (*jerryx_module_resolver_t) (const jerry_char_t *name, jerry_value_t *result);
+typedef struct
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name);
+  bool (*resolve) (const jerry_value_t canonical_name,
+                   jerry_value_t *result);
+};
```
Comment (LaszloLango) [inline docs/12.EXT-REFERENCE-MODULE.md]: `} jerryx_module_resolver_t;`

**Comment by gabrielschulhof** [comment]
@LaszloLango fixed.

File: docs/12.EXT-REFERENCE-MODULE.md [hunk 2]
```diff
@@ -130,20 +163,36 @@ load_and_evaluate_js_file (const jerry_char_t *name, jerry_value_t *result)
 
   return return_value;
 }
+
+static jerry_value_t
+canonicalize_file_path (const jerry_value_t name)
+{
+  jerry_value_t resulting_path;
+
+  /* Search name for "./" and "../", follow symlinks, etc., then create resulting_path via jerry_create_string () */
```
Comment (zherczeg) [inline docs/12.EXT-REFERENCE-MODULE.md]: Full stop after sentence.

File: docs/12.EXT-REFERENCE-MODULE.md [hunk 3]
```diff
@@ -130,20 +163,36 @@ load_and_evaluate_js_file (const jerry_char_t *name, jerry_value_t *result)
...
+};
 ```
 
 We can now load JavaScript files:
 ```c
 static const jerryx_module_resolver_t resolvers[] =
 {
   /* Consult the JerryScript module resolver first, in case the requested module is a compiled-in JerryScript module. */
-  jerryx_module_native_resolver,
+  &jerryx_module_native_resolver,
```
Comment (zherczeg) [inline docs/12.EXT-REFERENCE-MODULE.md]: Why do you need & ?

File: jerry-ext/module/module.c [hunk 4]
```diff
@@ -137,33 +131,52 @@ jerryx_module_add_to_cache (jerry_value_t cache, /**< cache to which to add the
...
+jerryx_resolve_native_module (const jerry_value_t name, /**< name of the module */
                                jerry_value_t *result) /**< [out] where to put the resulting module instance */
 {
   int index;
   const jerryx_native_module_t *module_p = NULL;
 
+  bool return_value = false;
+
+  jerry_size_t c_module_name_size = jerry_get_utf8_string_size (name);
+  jerry_char_t c_module_name[c_module_name_size];
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: While this allocation is simple in source code level, I always has doubts about its efficiency. You need to move a lot of data on the stack.

File: jerry-ext/module/module.c [hunk 5]
```diff
@@ -180,29 +193,60 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
...
+                       const jerry_size_t count) /**< length of list */
 {
   size_t index;
-  jerry_value_t ret;
-  jerry_value_t instances = *(jerry_value_t *) jerry_get_context_data (&jerryx_module_manager);
-  jerry_value_t module_name = jerry_create_string_from_utf8 (name);
+  jerry_value_t ret = 0;
+  const jerryx_module_resolver_t *resolver_p;
+  jerry_value_t instances;
+  jerry_value_t canonical_names_static[5];
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: Lets use a define for this.

File: jerry-ext/module/module.c [hunk 6]
```diff
@@ -180,29 +193,60 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
...
+    {
+      goto bail;
+    }
+  }
+
+  instances = *(jerry_value_t *) jerry_get_context_data (&jerryx_module_manager);
+
+  for (index = 0; index < count; index++)
+  {
+    canonical_names[index] = 0;
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: 0? Not undefined?

File: jerry-ext/module/module.c [hunk 7]
```diff
@@ -180,29 +193,60 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
...
 
   /* Return the cached instance if present. */
-  if (jerryx_module_check_cache (instances, module_name, &ret))
+  for (index = 0; index < count; index++)
   {
-    goto done;
+    resolver_p = resolvers_p[index];
+    canonical_names[index] =
+    resolver_p->get_canonical_name == NULL ? jerry_acquire_value (name) :
+    resolver_p->get_canonical_name (name);
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: Interesting alignment.

File: docs/12.EXT-REFERENCE-MODULE.md (hunk 3 above)
Comment (gabrielschulhof) [reply docs/12.EXT-REFERENCE-MODULE.md]: Because `jerryx_module_native_resolver` is now a structure stored in a global variable, and this is an array of structure pointers, so I will update the declaration of `resolvers` a few lines up.

File: jerry-ext/module/module.c (hunk 4 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: So, should I use `jmem_heap_alloc_block_null_on_error()` instead?

File: jerry-ext/module/module.c (hunk 5 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: What do you mean "use a define"?

File: jerry-ext/module/module.c (hunk 7 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: If I move the else under `jerry_acquire_value ()` I get this from check-vera.sh:
```
./jerry-ext/module/module.c:235: error: Indentation: 45 -> 4.
```
So, that's why it's like this.

File: jerry-ext/module/module.c (hunk 4 above)
Comment (zherczeg) [reply jerry-ext/module/module.c]: I am just asking what is its cost. We could introduce convenience stuff for them (I saw you already created some macros for it).

File: jerry-ext/module/module.c (hunk 5 above)
Comment (zherczeg) [reply jerry-ext/module/module.c]: `#define CANONICAL_NAMES_LOCAL_SIZE 4` or something. I think 4 is enough (probably 2 is the most frequent in practice).

File: jerry-ext/module/module.c (hunk 7 above)
Comment (zherczeg) [reply jerry-ext/module/module.c]: Use round brackets.

File: jerry-ext/module/module.c [hunk 8]
```diff
@@ -137,33 +131,52 @@ jerryx_module_add_to_cache (jerry_value_t cache, /**< cache to which to add the
...
 
-  return false;
-} /* jerryx_module_native_resolver */
+  return return_value;
+} /* jerryx_resolve_native_module */
+
+jerryx_module_resolver_t jerryx_module_native_resolver =
+{
+  NULL,
+  jerryx_resolve_native_module
```
Comment (martijnthe) [inline jerry-ext/module/module.c]:
```
.get_canonical_name = NULL,
.resolve = jerryx_resolve_native_module,
```

File: jerry-ext/include/jerryscript-ext/module.h [hunk 9]
```diff
@@ -22,6 +22,24 @@
...
 #endif /* __GNUC__ */
 
+/**
+ * Declare the structure for module resolvers.
+ */
+typedef struct
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name); /**< module name to canonicalize */
+  bool (*resolve) (const jerry_value_t canonical_name, /**< requested module's canonical name */
+                   jerry_value_t *result); /**< resulting module */
```
Comment (martijnthe) [inline jerry-ext/include/jerryscript-ext/module.h]: What's your rationale for passing around the name strings as `jerry_value_t`s?
I think most module resolvers will convert it right back to a C string.
Is it to avoid having to deal with memory management of the C strings?

Comment (gabrielschulhof) [reply jerry-ext/include/jerryscript-ext/module.h]: Yeah, and I was actually thinking that, for path resolution, you might not even convert the whole value to a C string, but you would rather use fixed size chunks to gradually retrieve the string and maybe assemble the canonical name in a similarly gradual fashion. Then, unless you had a directory with a really, really, really long name, you wouldn't really be allocating all that much.

Comment (gabrielschulhof) [reply jerry-ext/include/jerryscript-ext/module.h]: Call it streamed path resolution.

Comment (gabrielschulhof) [reply jerry-ext/include/jerryscript-ext/module.h]: Another very important reason: the module cache is stored as a JavaScript object, so I have to convert a C string back to a `jerry_value_t` in order to check if that JavaScript object has a module at the module name's key.

Comment (martijnthe) [reply jerry-ext/include/jerryscript-ext/module.h]: Ah I see, yeah that makes sense.

**Comment by gabrielschulhof** [comment]
@martijnthe @LaszloLango @zherczeg I have updated the patch to reflect the changes to the moduling system.

File: jerry-ext/include/jerryscript-ext/module.h [hunk 10]
```diff
@@ -92,20 +92,27 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
...
  */
-typedef bool (*jerryx_module_resolver_t) (const jerry_char_t *name, jerry_value_t *result);
+extern jerryx_module_resolver_t jerryx_module_native_resolver;
 
 /**
  * Load a copy of a module into the current context using the provided module resolvers, or return one that was already
  * loaded if it is found.
  */
-jerry_value_t jerryx_module_resolve (const jerry_char_t *name, const jerryx_module_resolver_t *resolvers, size_t count);
+jerry_value_t jerryx_module_resolve (const jerry_value_t name, /**< module's name */
```
Comment (zherczeg) [inline jerry-ext/include/jerryscript-ext/module.h]: Ditto

File: jerry-ext/include/jerryscript-ext/module.h [hunk 11]
```diff
@@ -92,20 +92,27 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
...
- * Declare the JerryScript module resolver so that it may be added to an array of jerryx_module_resolver_t items and
- * thus passed to jerryx_module_resolve.
+ * Declare the structure for module resolvers.
  */
-bool jerryx_module_native_resolver (const jerry_char_t *name, jerry_value_t *result);
+typedef struct
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name);  /**< module name to canonicalize */
+  bool (*resolve) (const jerry_value_t canonical_name,  /**< requested module's canonical name */
+                   jerry_value_t *result); /**< resulting module */
```
Comment (zherczeg) [inline jerry-ext/include/jerryscript-ext/module.h]: We do not use comments in header files.

File: jerry-ext/module/module.c [hunk 12]
```diff
@@ -167,26 +167,38 @@ static const jerry_char_t *on_resolve_absent = (jerry_char_t *) "Module on_resol
...
-bool
-jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module */
-                               jerry_value_t *result) /**< [out] where to put the resulting module instance */
+static bool
+jerryx_resolve_native_module (const jerry_value_t canonical_name, /**< canonical name of the module */
+                              jerry_value_t *result) /**< [out] where to put the resulting module instance */
 {
   const jerryx_native_module_t *module_p = NULL;
 
+  jerry_size_t name_size = jerry_get_utf8_string_size (canonical_name);
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: Shall we call tostring on the name? Or check if it is a string.

File: jerry-ext/module/module.c [hunk 13]
```diff
@@ -202,29 +214,51 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
  *   - an error indicating that something went wrong during the attempt to load the module.
  */
 jerry_value_t
-jerryx_module_resolve (const jerry_char_t *name, /**< name of the module to load */
-                       const jerryx_module_resolver_t *resolvers_p, /**< list of resolvers */
+jerryx_module_resolve (const jerry_value_t name, /**< name of the module to load */
+                       const jerryx_module_resolver_t **resolvers_p, /**< list of resolvers */
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: `**` ? A pointer to a list of pointers which points to the resolvers? Why not an array of resolvers?

File: jerry-ext/module/module.c [hunk 14]
```diff
@@ -202,29 +214,51 @@ jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module
...
+  /**
+   * Establish the canonical name for the requested module. Each resolver presents its own canonical name. If one of
+   * the canonical names matches a cached module, it is returned as the result.
+   */
+  for (index = 0; index < resolver_count; index++)
+  {
+    resolver_p = resolvers_p[index];
+    canonical_names[index] = ((resolver_p && resolver_p->get_canonical_name != NULL) ?
+      resolver_p->get_canonical_name (name):
+      jerry_acquire_value (name));
```
Comment (zherczeg) [inline jerry-ext/module/module.c]: This is not how we align ?:

File: tests/unit-ext/module/jerry-module-test.c [hunk 15]
```diff
@@ -61,41 +61,61 @@ const char eval_string5[] =
 "}) ();";
 
 static bool
-resolve_differently_handled_module (const jerry_char_t *name,
+resolve_differently_handled_module (const jerry_value_t name,
                                                 jerry_value_t *result)
```
Comment (zherczeg) [inline tests/unit-ext/module/jerry-module-test.c]: Comment is missing, spaces are wrong.

File: jerry-ext/module/module.c (hunk 12 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: We shouldn't need to. This function should not be called directly, even though a pointer to it is available. It should only be called by `jerryx_module_resolve ()` and it should only ever receive a `jerry_value_t` stored in the array of canonical module names created by `jerryx_module_resolve ()`. In fact, this function is not even documented. Only the structure `jerryx_module_native_resolver` is documented, and only the availability of it as a global data symbol is documented, not its contents.

File: jerry-ext/module/module.c (hunk 13 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: Because the resolvers are stored in-contiguously, and we do not want to make copies of them, right?

File: jerry-ext/module/module.c (hunk 14 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: It's pretty tough not breaking 120 columns with this line. I'm still trying to figure out how to align it.

Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: I renamed `resolver_p` to `res_p` then it fits.

**Comment by gabrielschulhof** [comment]
@zherczeg I have addressed your review comments.

File: jerry-ext/module/module.c (hunk 13 above)
Comment (zherczeg) [reply jerry-ext/module/module.c]: Are they used multiple times? Or why they are scattered?

I think one application use one resolver list, but this is not much of a code size increase, so lleave it be.

File: jerry-ext/module/module.c (hunk 14 above)
Comment (zherczeg) [reply jerry-ext/module/module.c]: I would move the condition into a bool value rather than shoretning the name.

File: jerry-ext/module/module.c (hunk 13 above)
Comment (gabrielschulhof) [reply jerry-ext/module/module.c]: Resolvers may be declared global in separate .c files (jerryx_module_native_resolver` certainly is), and they are collected into an array only once - the place that calls `jerryx_module_resolve ()`.

**Comment by gabrielschulhof** [comment]
@zherczeg instead of a boolean I went straight for storing the function pointers in variables.

**Comment by gabrielschulhof** [comment]
... aaand I forgot to key the cache on the canonical name. Fixed now.

**Comment by gabrielschulhof** [comment]
@zherczeg could you or somebody with Travis access please restart the lint job? It seems to have been killed because it had stalled.

**Comment by gabrielschulhof** [comment]
Thanks!

File: jerry-ext/include/jerryscript-ext/module.h [hunk 16]
```diff
@@ -29,9 +29,9 @@ typedef jerry_value_t (*jerryx_native_module_on_resolve_t) (void);
  */
 typedef struct jerryx_native_module_t
 {
-  const jerry_char_t *name_p; /**< name of the module */
-  const jerryx_native_module_on_resolve_t on_resolve; /**< function that returns a new instance of the module */
-  struct jerryx_native_module_t *next_p; /**< pointer to next module in the list */
+  const jerry_char_t *name_p;
+  const jerryx_native_module_on_resolve_t on_resolve;
+  struct jerryx_native_module_t *next_p;
```
Comment (LaszloLango) [inline jerry-ext/include/jerryscript-ext/module.h]: Why did you remove the comments from here?

File: jerry-ext/module/module.c [hunk 17]
```diff
@@ -14,11 +14,13 @@
  */
 
 #include <string.h>
+#include "jmem.h"
```
Comment (LaszloLango) [inline jerry-ext/module/module.c]: I think we should not use internal headers in `jerry-ext`. I don't see what are you using from it. Please remove it if possible.

File: jerry-ext/include/jerryscript-ext/module.h [hunk 18]
```diff
@@ -92,20 +92,27 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
...
 
 /**
- * Declare the JerryScript module resolver so that it may be added to an array of jerryx_module_resolver_t items and
- * thus passed to jerryx_module_resolve.
+ * Declare the structure for module resolvers.
  */
-bool jerryx_module_native_resolver (const jerry_char_t *name, jerry_value_t *result);
+typedef struct
+{
+  jerry_value_t (*get_canonical_name) (const jerry_value_t name);
```
Comment (LaszloLango) [inline jerry-ext/include/jerryscript-ext/module.h]: Missing comments

File: jerry-ext/include/jerryscript-ext/module.h (hunk 16 above)
Comment (gabrielschulhof) [reply jerry-ext/include/jerryscript-ext/module.h]: @zherczeg [said](https://github.com/jerryscript-project/jerryscript/pull/2013#discussion_r140710783) we didn't add such `/**<` comments to the header files. Or did I misinterpret?

File: jerry-ext/include/jerryscript-ext/module.h (hunk 18 above)
Comment (gabrielschulhof) [reply jerry-ext/include/jerryscript-ext/module.h]: I need to know for sure. Do we, or do we not add `/**<` comments after structure members and function arguments? How do we document function pointer structure members, since the formal parameters of the function signature also needs to be documented?

File: jerry-ext/include/jerryscript-ext/module.h (hunk 16 above)
Comment (LaszloLango) [reply jerry-ext/include/jerryscript-ext/module.h]: Yes, I think there was a misunderstanding here. We do not add comments to declarations, but only to definitions. The type definitions are usually in the header files, so we must put the comments there. Please don't remove these comments.

Comment (zherczeg) [reply jerry-ext/include/jerryscript-ext/module.h]: Sorry, I mean only for function arguments, not for types.

**Comment by gabrielschulhof** [comment]
@LaszloLango fixed up the comments and added some `_p`s.

File: jerry-ext/include/jerryscript-ext/module.h [hunk 19]
```diff
@@ -92,20 +92,41 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
 void jerryx_native_module_unregister (jerryx_native_module_t *module_p);
 
 /**
- * Declare the JerryScript module resolver so that it may be added to an array of jerryx_module_resolver_t items and
- * thus passed to jerryx_module_resolve.
+ * Declare the function pointer type for canonical name resolution.
+ */
+typedef jerry_value_t (*jerryx_module_get_canonical_name_t) (const jerry_value_t name); /**< The name for which to
+                                                                                         * compute the canonical
+                                                                                         * name */
```
Comment (zherczeg) [inline jerry-ext/include/jerryscript-ext/module.h]: Align the start of the comment lines to the same column.

File: jerry-ext/include/jerryscript-ext/module.h [hunk 20]
```diff
@@ -92,20 +92,41 @@ void jerryx_native_module_register (jerryx_native_module_t *module_p);
...
+
+/**
+ * Declare the function pointer type for module resolution.
  */
-bool jerryx_module_native_resolver (const jerry_char_t *name, jerry_value_t *result);
+typedef bool (*jerryx_module_resolve_t) (const jerry_value_t canonical_name, /**< The module's canonical name */
+                                         jerry_value_t *result); /**< The resulting module, if the function returns
+                                                                  * true */
```
Comment (zherczeg) [inline jerry-ext/include/jerryscript-ext/module.h]: Ditto. And similar to below.

File: tests/unit-ext/test-ext-module-canonical.c [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+get_canonical_name (const jerry_value_t name)
+{
+  jerry_size_t name_size = jerry_get_string_size (name);
+  jerry_char_t name_string[name_size + 1];
+  jerry_string_to_char_buffer (name, name_string, name_size);
+  name_string[name_size] = 0;
+
+  return (!strcmp ((char *) name_string, "batman") ? jerry_acquire_value (name):
+         (!strcmp ((char *) name_string, "bruce-wayne") ? jerry_create_string ((jerry_char_t *) "batman"):
+         jerry_create_undefined ()));
```
Comment (zherczeg) [inline tests/unit-ext/test-ext-module-canonical.c]: Fix the style of these lines.

File: tests/unit-ext/test-ext-module-canonical.c [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+{
+  (void) argc;
+  (void) argv;
+
+  const jerryx_module_resolver_t *resolver = &canonical_test;
+
+  jerry_init (JERRY_INIT_EMPTY);
+
+  jerry_value_t batman = jerry_create_string ((jerry_char_t *) "batman");
+  jerry_value_t bruce_wayne = jerry_create_string ((jerry_char_t *) "bruce-wayne");
```
Comment (zherczeg) [inline tests/unit-ext/test-ext-module-canonical.c]: I hope these are not copyrighted.

Comment (gabrielschulhof) [reply tests/unit-ext/test-ext-module-canonical.c]: Good point. I'll use [Liberator](http://pdsh.wikia.com/wiki/Liberator), which is public domain.

Comment (zherczeg) [reply tests/unit-ext/test-ext-module-canonical.c]: They are still, however, "use at your own risk" characters. - this doesn't sound entirely public

Comment (gabrielschulhof) [reply tests/unit-ext/test-ext-module-canonical.c]: *sigh* Alice and Bob it is.

**Comment by gabrielschulhof** [comment]
@zherczeg @LaszloLango please have another look. I updated the test to use `#define`-ed strings to store the alias and the actual name at the top, and I replaced the ternary composition with `if`/`else if`/`else`. I also changed the resolver initializer to use the `{ .member_name = value }` notation.

**Comment by LaszloLango** [comment]
still LGTM

**Comment by zherczeg** [comment]
Still ok.
